This patch is based on the commit described below, taken from upstream pull
request 3175.

It was adjusted by Peter Michael Green to use target_endian instead of hard
coding big and little endian architectures.

commit 7f3a1e7727377045d73daca272711c548fe0f394
Author: Ola x Nilsson <olani@axis.com>
Date:   Mon Mar 27 14:17:19 2023 +0200

    Update struct timespec for _TIME_BITS=64
    
    powerpc and mips wants 32 bits of padding before tv_nsec.
    arm and x86 wants 32bits of padding after tv_nsec.

diff --git a/src/unix/mod.rs b/src/unix/mod.rs
index 42a293e9b0..f120933db6 100644
--- a/src/unix/mod.rs
+++ b/src/unix/mod.rs
@@ -74,10 +74,14 @@ s! {
     // See https://sourceware.org/bugzilla/show_bug.cgi?id=16437
     pub struct timespec {
         pub tv_sec: time_t,
+        #[cfg(all(gnu_time64_abi, target_endian = "big"))]
+        __pad: i32,
         #[cfg(all(target_arch = "x86_64", target_pointer_width = "32"))]
         pub tv_nsec: i64,
         #[cfg(not(all(target_arch = "x86_64", target_pointer_width = "32")))]
         pub tv_nsec: ::c_long,
+        #[cfg(all(gnu_time64_abi, target_endian = "little"))]
+        __pad: i32,
     }
 
     pub struct rlimit {
